From 3dc731657a721a48344a443ce628eac73da0b4b8 Mon Sep 17 00:00:00 2001
From: MariuszSzczepanikSpyrosoft
 <118888269+MariuszSzczepanikSpyrosoft@users.noreply.github.com>
Date: Wed, 14 Jun 2023 16:31:18 +0200
Subject: [PATCH] Fix build according to modification of GPSD (#78)

GPSD since version 3.23.1 (GPSD_API_MAJOR_VERSION=12) has introduced
interfaces changes on the below values:
- Change STATUS_NO_FIX to STATUS_UNK to avoid confusion with fix mode.
- Change STATUS_FIX to STATUS_GPS to avoid confusion with fix mode.
- Change STATUS_DGPS_FIX to STATUS_DGPS to avoid confusion with fix
mode.
---
 gpsd_client/src/client.cpp | 29 +++++++++++++++++++----------
 1 file changed, 19 insertions(+), 10 deletions(-)

diff --git a/src/client.cpp b/src/client.cpp
index 579f0e8..736bfef 100644
--- a/src/client.cpp
+++ b/src/client.cpp
@@ -148,7 +148,9 @@ class GPSDClient {
 #endif
       }
 
-#if GPSD_API_MAJOR_VERSION >= 10
+#if GPSD_API_MAJOR_VERSION >= 12
+      if ((p->fix.status & STATUS_GPS) && !(check_fix_by_variance && std::isnan(p->fix.epx))) {
+#elif GPSD_API_MAJOR_VERSION >= 10
       if ((p->fix.status & STATUS_FIX) && !(check_fix_by_variance && std::isnan(p->fix.epx))) {
 #else
       if ((p->status & STATUS_FIX) && !(check_fix_by_variance && std::isnan(p->fix.epx))) {
@@ -157,9 +159,11 @@ class GPSDClient {
         status.status = 0; // FIXME: gpsmm puts its constants in the global
                            // namespace, so `GPSStatus::STATUS_FIX' is illegal.
 
-// STATUS_DGPS_FIX was removed in API version 6 but re-added afterward
+// STATUS_DGPS_FIX was removed in API version 6 but re-added afterward and next renamed since the version 12
 #if GPSD_API_MAJOR_VERSION != 6
-#if GPSD_API_MAJOR_VERSION >= 10
+#if GPSD_API_MAJOR_VERSION >= 12
+        if (p->fix.status & STATUS_DGPS)
+#elif GPSD_API_MAJOR_VERSION >= 10
         if (p->fix.status & STATUS_DGPS_FIX)
 #else
         if (p->status & STATUS_DGPS_FIX)
@@ -206,7 +210,7 @@ class GPSDClient {
 
         /* TODO: attitude */
       } else {
-        status.status = -1; // STATUS_NO_FIX
+        status.status = -1; // STATUS_NO_FIX or STATUS_UNK
       }
 
       fix.status = status;
@@ -242,15 +246,20 @@ class GPSDClient {
 #else
       switch (p->status) {
 #endif
+#if GPSD_API_MAJOR_VERSION >= 12
+        case STATUS_GPS:
+#else
         case STATUS_NO_FIX:
-          fix->status.status = -1; // NavSatStatus::STATUS_NO_FIX;
-          break;
-        case STATUS_FIX:
-          fix->status.status = 0; // NavSatStatus::STATUS_FIX;
+#endif
+          fix->status.status = 0; // NavSatStatus::STATUS_FIX or NavSatStatus::STATUS_GPS;
           break;
-// STATUS_DGPS_FIX was removed in API version 6 but re-added afterward
-#if GPSD_API_MAJOR_VERSION != 6 
+// STATUS_DGPS_FIX was removed in API version 6 but re-added afterward and next renamed since the version 12
+#if GPSD_API_MAJOR_VERSION != 6
+#if GPSD_API_MAJOR_VERSION >= 12
+        case STATUS_DGPS:
+#else
         case STATUS_DGPS_FIX:
+#endif
           fix->status.status = 2; // NavSatStatus::STATUS_GBAS_FIX;
           break;
 #endif
